modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/near/IgniteTxExceptionNodeFailTest.java

```diff
@@ -0,0 +1,193 @@
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.apache.ignite.transactions.Transaction;
+import org.apache.ignite.transactions.TransactionHeuristicException;
+import org.junit.Test;
+import org.locationtech.jts.util.Assert;
+
+import static org.apache.ignite.cache.CacheWriteSynchronizationMode.FULL_SYNC;
+import static org.apache.ignite.cache.CacheWriteSynchronizationMode.PRIMARY_SYNC;
+import static org.apache.ignite.internal.TestRecordingCommunicationSpi.spi;
+
+/**
+ * Tests check a result of commit when a node fail before
+ * send {@link GridNearTxFinishResponse} to transaction coodinator
+ */
+public class IgniteTxExceptionNodeFailTest extends GridCommonAbstractTest {
+    /** syncMode */
+    private static CacheWriteSynchronizationMode syncMode;
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        DataStorageConfiguration dsConfig = new DataStorageConfiguration()
+            .setDefaultDataRegionConfiguration(new DataRegionConfiguration()
+                .setPersistenceEnabled(true));
+
+        cfg.setCommunicationSpi(new TestRecordingCommunicationSpi());
+
+        return cfg
+            .setDataStorageConfiguration(dsConfig)
+            .setCacheConfiguration(new CacheConfiguration("cache")
+                .setAtomicityMode(CacheAtomicityMode.TRANSACTIONAL)
+                .setWriteSynchronizationMode(syncMode).setBackups(0));
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        FileUtils.deleteDirectory(new File(U.defaultWorkDirectory()));
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        stopAllGrids();
+    }
+
+    /**
+     * Test with {@link CacheWriteSynchronizationMode#PRIMARY_SYNC}
+     * @throws Exception
+     */
+    @Test
+    public void testNodeFailWithPrimarySync() throws Exception {
+        testNodeFail(PRIMARY_SYNC);
+    }
+
+    /**
+     * Test with {@link CacheWriteSynchronizationMode#FULL_SYNC}
+     * @throws Exception
+     */
+    @Test
+    public void testNodeFailWithFullSync() throws Exception {
+        testNodeFail(FULL_SYNC);
+    }
+
+    /**
+     * <ul>
+     * <li>Start 2 nodes with transactional cache without backups
+     * <li>Start transaction:
+     *  <ul>
+     *  <li>put a key to a partition on transaction coordinator
+     *  <li>put a key to a partition on other node
+     *  <li>try to commit the transaction
+     *  </ul>
+     * <li>Stop other node when it try to send GridNearTxFinishResponse
+     * <li>Check that {@link Transaction#commit()} throw {@link TransactionHeuristicException}
+     * </ul>
+     *
+     * @param testSyncMode
```

Comment:
Please add description for parameter and exception class. IDEA highlights absence of a comment as a warning. Applicable to several more places in the test class.

Reply:
![image](https://user-images.githubusercontent.com/2736390/86587177-d866cb80-bf91-11ea-97d0-cb685fbc57c6.png)

Reply:
fixed